File: Maximum Depth of N-ary Tree/Maximum Depth of N-ary Tree.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public int maxDepth(Node root) {
        int maxDept=0;
        return dfs(root,maxDept);
    }
    
    public int dfs(Node n, int dept) {
        if(n==null) return 0;  // once reaching a node who has no childrens it goes back to the previous dfs call
            
        int max = 0;
        // for every children of the nodes call function again with recursion 
        // if I have 3 children node I set the new max to the node who has the most descendents
        for(Node child: n.children) {
            // for example my childrens are node 5 and node 6
            // maxDepth call for node 5 return value is return value of dfs is 1
            // maxDepth call for node 6 return value is return value of dfs is 1
            max = Math.max(max, maxDepth(child));
        }
        // starting off the bottom of the tree every time a node goes back to
        // its parent, increase depth by 1 
        return 1 + max;
    }
    
}

File: Maximum Depth of N-ary Tree/Maximum Depth of N-ary Tree.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:

    def maxDepth(self, root: 'Node') -> int:
        return self.dfs(root) # pass the root node
    
    def dfs(self,node:'Node')->int:
        if node==None:              # if the current passed node is null then we know
           return 0                 # his parent is a leaf node
        
        depth=0                     # will be set by the loop
                                    # ex: child1:2 child2:1 child3:3    depth=3 
        for i in node.children:
           depth=max(depth,self.maxDepth(i))  # a given node I check which of the children has more depth
                                              # like finding max in array we call maxDepth to indirectly call dfs for child nodes
        return 1+depth              # current node adds a level to its child's depth so 3+1
        

        # when function reaches the child of a leaf node aka None, None Node will return 0 in self.maxDepth
        # the first leaf node will compare max(depth,0) they both are 0 so depth=0
        # when leaf node return his depth to his parent node we increase of 1 level so depth+1 (0+1)
        # parent node compares this depth with other child to see which is higher
        # then overwrite his own depth and adds 1 to it for his own parent